packages/bns/src/index.ts

```diff
@@ -0,0 +1,764 @@
+      } else {
+        throw new Error('Response did not contain a number');
+      }
+    } else {
+      const errorResponse = responseCV as ResponseErrorCV;
+      throw new Error(cvToString(errorResponse.value));
+    }
+  });
+}
+
+/**
+ * Get price of name registration in microstacks
+ *
+ * @param {string} fullyQualifiedName - the fully qualified name
+ * @param {StacksNetwork} network - the Stacks network to use
+ *
+ * @returns {Promise} that resolves to a BN object number of microstacks if the operation succeeds
+ */
+export async function getNamePrice(
+  fullyQualifiedName: string,
+  network?: StacksNetwork
+): Promise<BN> {
+  const bnsFunctionName = 'get-name-price';
+  const { subdomain, namespace, name } = decodeFQN(fullyQualifiedName);
+  if (subdomain) {
+    throw new Error('Cannot get subdomain name price');
+  }
+
+  // Create a random address as input to read-only function call
+  // Not used by BNS contract function but required by core node API
+  // https://github.com/blockstack/stacks-blockchain/blob/master/src/net/http.rs#L1796
+  const randomPrivateKey = privateKeyToString(makeRandomPrivKey());
+  const randomAddress = getAddressFromPrivateKey(randomPrivateKey);
+
+  return callReadOnlyBNSFunction({
+    functionName: bnsFunctionName,
+    senderAddress: randomAddress,
+    functionArgs: [bufferCVFromString(namespace), bufferCVFromString(name)],
+    network: network || new StacksMainnet(),
+  }).then((responseCV: ClarityValue) => {
+    if (responseCV.type === ClarityType.ResponseOk) {
+      if (responseCV.value.type === ClarityType.Int || responseCV.value.type === ClarityType.UInt) {
+        return responseCV.value.value;
+      } else {
+        throw new Error('Response did not contain a number');
+      }
+    } else {
+      const errorResponse = responseCV as ResponseErrorCV;
+      throw new Error(cvToString(errorResponse.value));
+    }
+  });
+}
+
+/**
+ * Preorder namespace options
+ *
+ * @param  {String} namespace - the namespace to preorder
+ * @param  {String} salt - salt used to generate the preorder namespace hash
+ * @param  {BigNum} stxToBurn - amount of STX to burn for the registration
+ * @param  {String} publicKey - the private key to sign the transaction
+ * @param  {StacksNetwork} network - the Stacks blockchain network to register on
+ */
+export interface PreorderNamespaceOptions {
+  namespace: string;
+  salt: string;
+  stxToBurn: BN;
+  publicKey: string;
+  network?: StacksNetwork;
+}
+
+/**
+ * Generates a namespace preorder transaction.
+ * First step in registering a namespace. This transaction does not reveal the namespace that is
+ * about to be registered. And it sets the amount of STX to be burned for the registration.
+ *
+ * Resolves to the generated StacksTransaction
+ *
+ * @param  {PreorderNamespaceOptions} options - an options object for the preorder
+ *
+ * @return {Promise<StacksTransaction>}
+ */
+export async function buildPreorderNamespaceTX({
```

Comment:
Why uppercase X, it's not an acronym.

```suggestion
export async function buildPreorderNamespaceTx({
```